File: MedalTable.py
class MedalTable:
	class Country:
		def __init__(self, name):
			self.name = name
			self.gold = 0
			self.silver = 0
			self.bronze = 0
		
		def __lt__(self, other):
			if self.gold != other.gold:
				return self.gold > other.gold
			if self.silver != other.silver:
				return self.silver > other.silver
			if self.bronze != other.bronze:
				return self.bronze > other.bronze
			return self.name < other.name
			
	def generate(self, results):
		countryDict = {}
		for result in results:
			strs = result.split(' ')
			for s in strs:
				if s not in countryDict:
					countryDict[s] = self.Country(s)
			countryDict[strs[0]].gold += 1
			countryDict[strs[1]].silver += 1
			countryDict[strs[2]].bronze += 1
		countries = countryDict.values()
		countries.sort()
		result = []
		for country in countries:
			result.append('%s %d %d %d' % (country.name, country.gold, country.silver, country.bronze))
		return result

File: tutorials/how_to_find_a_solution/MedalTable.cpp
// SRM 209 Div 1, Level One
#include <string>
#include <vector>
#include <unordered_map>
#include <algorithm>

using namespace std;

class MedalTable {
	struct Node {
		string name;
		int gold;
		int silver;
		int bronze;
		
		Node() : gold(0), silver(0), bronze(0) {}
	};
	
	static bool compareNode(const Node& n1, const Node& n2) {
		if (n1.gold != n2.gold) {
			return n1.gold > n2.gold;
		}
		if (n1.silver != n2.silver) {
			return n1.silver > n2.silver;
		}
		if (n1.bronze != n2.bronze) {
			return n1.bronze > n2.bronze;
		}
		return n1.name < n2.name;
	}

public:
	vector<string> generate(vector<string> results) {
		unordered_map<string, Node> map;
		for (auto& res : results) {
			char name[3][4];
			sscanf(res.c_str(), "%s %s %s", name[0], name[1], name[2]);
			for (int i = 0; i < 3; ++i) {
				auto it = map.find(name[i]);
				if (it == map.end()) {
					Node node;
					node.name = name[i];
					auto pair = map.insert(make_pair(node.name, node));
					it = pair.first;
				}
				if (i == 0) it->second.gold++;
				else if (i == 1) it->second.silver++;
				else if (i == 2) it->second.bronze++;
			}
		}
		vector<Node> nodes;
		for (auto pair : map) {
			nodes.push_back(pair.second);
		}
		sort(nodes.begin(), nodes.end(), MedalTable::compareNode);
		vector<string> rank;
		for (auto& node : nodes) {
			char buf[40];
			snprintf(buf, sizeof(buf), "%s %d %d %d", node.name.c_str(), node.gold, node.silver, node.bronze);
			rank.push_back(buf);
		}
		return rank;
	}
};
